Fix #8314: CLI PrimeFlex migration fixes

melloware opened the pull request:
Added unit tests proving the issue and fixed them

melloware committed:
Fix #8314: CLI PrimeFlex migration fixes

jepsar commented on primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:
```diff
@@ -57,8 +57,9 @@ protected void initReplaceRegEx() {
+        replaceRegex.put("p-offset-([0-9]+?)", "col-offset-$1");
         replaceRegex.put("p-(xl|lg|md|sm)-offset-([0-9]+?)", "$1:col-offset-$2");
-        replaceRegex.put("p-col", "col");
+        replaceRegex.put("p-col\s", "col\s");
```
What about `class="whatever p-col"`?

melloware replied:
ahhh crap your are right.  I was always assuming it was first or in the middle not on the end.

Any suggestions?

melloware replied:
Maybe `(\s|\")` do a capture group of space or `"` ?

jepsar replied:
Something like
```js
"<p-fieldset class=\"p-fieldset\"".replace(/([^<])p-fieldset/, '$1fieldset')
```
(JavaScript)

melloware commented:
OK this seems to fix it...

```java
replaceRegex.put("p-col(\s|\")", "col$1");
```

jepsar commented on primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:
```diff
@@ -57,8 +57,9 @@ protected void initReplaceRegEx() {
+        replaceRegex.put("p-offset-([0-9]+?)", "col-offset-$1");
         replaceRegex.put("p-(xl|lg|md|sm)-offset-([0-9]+?)", "$1:col-offset-$2");
-        replaceRegex.put("p-col", "col");
+        replaceRegex.put("p-col(\s|\")", "col$1");
```
What about `class='p-col'`?

melloware replied:
Good catch fixed single quote as well and added test case

melloware commented:
I don't understand why I am getting this on GitHub Actions when it compiles and works fine for me locally.

```
Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/primefaces/primefaces/primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:[62,34] illegal escape character
Error:  /home/runner/work/primefaces/primefaces/primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:[195,36] illegal escape character
```

jepsar commented:
Isn't it better to try to skip `<p-xxx` and `</p-xxx`? What if one would have `styleClass="p-col#{condtion ? ' p-class' : ''}"`?

ThoSap commented:
> Isn't it better to try to skip `<p-xxx` and `</p-xxx`? What if one would have `styleClass="p-col#{condtion ? ' p-class' : ''}"`?

Yes that this is better, but the regex must not only exclude occurrences starting with `<`  as `p-col` does not _only_ exist in HTML element opening and closing tags in PrimeNG.
The term `p-col` can also exist in CSS/SCSS/TS files without `<` as my migration yesterday showed.

For some examples see
https://github.com/primefaces/primefaces/issues/8314#issuecomment-1019341626

Some files you could test against:
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/colorpicker/colorpicker.css
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/colorpicker/colorpicker.ts#L16-L32
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/colorpicker/colorpicker.spec.ts

https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/table/table.css#L245-L311
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/table/table.ts#L4300-L4343
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/table/table.spec.ts

https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/fieldset/fieldset.css
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/fieldset/fieldset.ts#L11-L87
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/fieldset/fieldset.spec.ts#L28-L48

Occurrences with `p-col` and `p-col-12` which should still work after a regex change:
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/dataview/dataview.ts#L31
https://github.com/primefaces/primeng/blob/13.1.0/src/app/components/table/table.spec.ts#L236-L249

ThoSap commented:
Can't we use something like this with a negative lookahead?
For `p-col` https://regexr.com/6drum
For `p-field` https://regexr.com/6drvb

Lucky for us PrimeFlex 2.0.0 `p-col` and `p-field` subclasses always used a **-** inbetween and not a \w character (example `p-col-offset-7` instead of `p-coloffset-7`) 😆

ThoSap commented on primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:
```diff
@@ -133,6 +143,11 @@ protected void initReplaceRegEx() {
         replaceRegex.put("p-as-center", "align-self-center");
         replaceRegex.put("p-as-baseline", "align-self-baseline");
         replaceRegex.put("p-as-stretch", "align-self-stretch");
+        replaceRegex.put("p-col-align-start", "align-self-start");
```
Regarding my last comment https://github.com/primefaces/primefaces/pull/8315/files#issuecomment-1019457534, then you probably have to change the order of the specific `p-col-*` map entries to be first and the new negative lookahead regex must be last.

Otherwise for example `p-col-align-start` will be migrated to `col-align-start` instead of `align-self-start`.

ThoSap commented on primefaces-cli/src/main/java/org/primefaces/cli/migration/primeflex/PrimeFlexMigration.java:
```diff
@@ -176,7 +192,7 @@ protected void initReplaceRegEx() {
+        replaceRegex.put("p-field(\s|'|\")", "field$1");
         replaceRegex.put("p-formgrid", "formgrid");
         replaceRegex.put("p-formgroup-inline", "formgroup-inline");
         replaceRegex.put("p-field-checkbox", "field-checkbox");
```
To be consistent the same goes for specific `p-field-*` classes, even though the migration here would be correct (`p-field-checkbox` gets `field-checkbox`).

melloware replied:
I moved both lookahead regexes to the very bottom so they are checked last.

melloware commented:
@ThoSap I changed the `order` the way you suggested as well as switch to your two lookahead regexes they work like a charm!

BTW I use Regex 101 I like it a lot better only because it has a "Code Generator" option so it generated me the Java version of your regex. See https://regex101.com/r/lbCIQ9/1

Appreciate all your help on this!  I converted a PrimeReact project no problem but it was using a newer version of PrimeFlex 2.0.0 so I never ran into some of these issues but this will be a huge help to the whole PRIME community!
